Quartz API fails startup sometimes due to gRPC dependency

## Describe the bug

The container starts successfully and Gunicorn begins serving, but the FastAPI lifespan startup hook crashes while attempting to communicate with a downstream gRPC Data Platform service.

## Expected behavior
API should have error handling and a retry logic 

## Additional context

```
----------------------------------------
/var/log/eb-docker/containers/eb-current-app/eb-stdouterr.log
----------------------------------------
time="2026-05-21T00:08:17Z" level=warning msg="/var/app/current/docker-compose.yml: `version` is obsolete"
[2Kquartz-api  | /opt/app/.venv/lib/python3.12/site-packages/auth0_api_python/api_client.py:7: AuthlibDeprecationWarning: authlib.jose module is deprecated, please use joserfc instead.
[2Kquartz-api  | It will be compatible before version 2.0.0.
[2Kquartz-api  |   from authlib.jose import JsonWebKey, JsonWebToken
[2Kquartz-api  | [2026-05-21 00:08:20 +0000] [1] [INFO] Starting gunicorn 26.0.0
[2Kquartz-api  | [2026-05-21 00:08:20 +0000] [1] [INFO] Listening at: http://0.0.0.0:80 (1)
[2Kquartz-api  | [2026-05-21 00:08:20 +0000] [1] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2Kquartz-api  | [2026-05-21 00:08:20 +0000] [10] [INFO] Booting worker with pid: 10
[2Kquartz-api  | [2026-05-21 00:08:20 +0000] [1] [INFO] Control socket listening at /root/.gunicorn/gunicorn.ctl
[2Kquartz-api  | /opt/app/.venv/lib/python3.12/site-packages/auth0_api_python/api_client.py:7: AuthlibDeprecationWarning: authlib.jose module is deprecated, please use joserfc instead.
[2Kquartz-api  | It will be compatible before version 2.0.0.
[2Kquartz-api  |   from authlib.jose import JsonWebKey, JsonWebToken
[2Kquartz-api  | You might want to install geopandas
[2Kquartz-api  | You might want to install geopandas
[2Kquartz-api  | {"ts": 1779322102, "level": "INFO", "logger": "uvicorn.error", "msg": "Started server process [10]", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322102, "level": "INFO", "logger": "uvicorn.error", "msg": "Waiting for application startup.", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322102, "level": "INFO", "logger": "httpx", "msg": "HTTP Request: GET http://testserver/openapi.json \"HTTP/1.1 500 Internal Server Error\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322102, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/sync \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322102, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/startup \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322112, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/sync \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322123, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/sync \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322133, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/sync \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322143, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/sync \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322153, "level": "DEBUG", "logger": "httpx", "msg": "HTTP Request: POST https://hub.apitally.io/v2/3275abf2-8063-4982-a11f-b109688dcaa7/development/sync \"HTTP/1.1 202 Accepted\"", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322162, "level": "ERROR", "logger": "uvicorn.error", "msg": "Traceback (most recent call last):\n  File \"/opt/app/.venv/lib/python3.12/site-packages/starlette/routing.py\", line 638, in lifespan\n    async with self.lifespan_context(app) as maybe_state:\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.12/contextlib.py\", line 210, in __aenter__\n    return await anext(self.gen)\n           ^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/app/.venv/lib/python3.12/site-packages/apitally/starlette.py\", line 377, in wrapped_lifespan\n    async with lifespan(app):\n               ^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.12/contextlib.py\", line 210, in __aenter__\n    return await anext(self.gen)\n           ^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/app/.venv/lib/python3.12/site-packages/fastapi/routing.py\", line 216, in merged_lifespan\n    async with original_context(app) as maybe_original_state:\n               ^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.12/contextlib.py\", line 210, in __aenter__\n    return await anext(self.gen)\n           ^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/app/.venv/lib/python3.12/site-packages/quartz_api/cmd/main.py\", line 253, in _lifespan\n    resp = await storage.get_locations(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/app/.venv/lib/python3.12/site-packages/quartz_api/internal/backends/dataplatform/client.py\", line 455, in get_locations\n    resp = await self.dpc.ListLocations(req)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/app/.venv/lib/python3.12/site-packages/grpc/aio/_interceptor.py\", line 474, in __await__\n    response = yield from call.__await__()\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/app/.venv/lib/python3.12/site-packages/grpc/aio/_call.py\", line 329, in __await__\n    raise _create_rpc_error(\ngrpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:\n\tstatus = StatusCode.UNAVAILABLE\n\tdetails = \"Socket closed\"\n\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_status:14, grpc_message:\"Socket closed\"}\"\n>\n", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322162, "level": "ERROR", "logger": "uvicorn.error", "msg": "Application startup failed. Exiting.", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | {"ts": 1779322162, "level": "INFO", "logger": "gunicorn.error", "msg": "Worker exiting (pid: 10)", "pid": 10, "trace_id": "unset"}
[2Kquartz-api  | Sentry is attempting to send 2 pending events
[2Kquartz-api  | Waiting up to 2 seconds
[2Kquartz-api  | Press Ctrl-C to quit
[2Kquartz-api  | [2026-05-21 00:09:23 +0000] [1] [ERROR] Worker (pid:10) exited with code 3.
[2Kquartz-api  | [2026-05-21 00:09:23 +0000] [1] [ERROR] Shutting down: Master
[2Kquartz-api  | [2026-05-21 00:09:23 +0000] [1] [ERROR] Reason: Worker failed to boot.
quartz-api exited with code 3
time="2026-05-21T00:09:33Z" level=warning msg="/var/app/current/docker-compose.yml: `version` is obsolete"
[2Kquartz-api  | [2026-05-21 00:09:23 +0000] [1] [ERROR] Worker (pid:10) exited with code 3.
[2Kquartz-api  | [2026-05-21 00:09:23 +0000] [1] [ERROR] Shutting down: Master
[2Kquartz-api  | [2026-05-21 00:09:23 +0000] [1] [ERROR] Reason: Worker failed to boot.
time="2026-05-21T00:09:43Z" level=warning msg="/var/app/current/docker-compose.yml: `version` is obsolete"
time="2026-05-21T00:09:54Z" level=warning msg="/var/app/current/docker-compose.yml: `version` is obsolete"
time="2026-05-21T00:10:04Z" level=warning msg="/var/app/current/docker-compose.yml: `version` is obsolete"
```

--- a/src/quartz_api/cmd/main.py
+++ b/src/quartz_api/cmd/main.py
@@ -274,7 +274,7 @@ async def _lifespan(server: FastAPI, conf: ConfigTree) -> AsyncGenerator[None]:
                     )
                     if attempt == 5:
                         raise
-                    await asyncio.sleep(5)
+                    await asyncio.sleep(1 * (2 ** (attempt - 1)))
 
         case _ as backend_type:
             raise ValueError(f"Unknown backend: {backend_type}")

--- a/src/quartz_api/cmd/main.py
+++ b/src/quartz_api/cmd/main.py
@@ -236,34 +236,45 @@ async def _lifespan(server: FastAPI, conf: ConfigTree) -> AsyncGenerator[None]:
             storage = DummyStorage()
             log.warning("disabled backend. NOT recommended for production")
         case "dataplatform":
-            from ocf.dp.dp_data import service_pb2_grpc
+            for attempt in range(1, 6):
+                try:
+                    from ocf.dp.dp_data import service_pb2_grpc
 
-            trace_interceptor = trace.TraceInterceptor()
-            grpc_channel = grpc.aio.insecure_channel(
-                target=conf.get_string("backend.dataplatform.host")
-                + ":"
-                + conf.get_string("backend.dataplatform.port"),
-                interceptors=[trace_interceptor],
-            )
-            client = service_pb2_grpc.DataPlatformDataServiceStub(grpc_channel)
-            storage = DataPlatformStorage.from_dp(dp_client=client)
+                    trace_interceptor = trace.TraceInterceptor()
+                    grpc_channel = grpc.aio.insecure_channel(
+                        target=conf.get_string("backend.dataplatform.host")
+                        + ":"
+                        + conf.get_string("backend.dataplatform.port"),
+                        interceptors=[trace_interceptor],
+                    )
+                    client = service_pb2_grpc.DataPlatformDataServiceStub(grpc_channel)
+                    storage = DataPlatformStorage.from_dp(dp_client=client)
 
-            if "uk_national" in conf.get_string("api.routers").split(","):
-                # Populate the GSP ID to UUID mapping
-                resp = await storage.get_locations(
-                    location_type=models.LocationType.GSP,
-                    energy_type=models.EnergyType.SOLAR,
-                    authdata={},
-                )
-                resp += await storage.get_locations(
-                    location_type=models.LocationType.NATION,
-                    energy_type=models.EnergyType.SOLAR,
-                    authdata={},
-                )
-                for loc in resp:
-                    if "gsp_id" in loc.metadata:
-                        gsp_id_map[int(loc.metadata["gsp_id"])] = loc
-                log.info(f"Populated GSP ID map with {len(gsp_id_map)} entries")
+                    if "uk_national" in conf.get_string("api.routers").split(","):
+                        # Populate the GSP ID to UUID mapping
+                        resp = await storage.get_locations(
+                            location_type=models.LocationType.GSP,
+                            energy_type=models.EnergyType.SOLAR,
+                            authdata={},
+                        )
+                        resp += await storage.get_locations(
+                            location_type=models.LocationType.NATION,
+                            energy_type=models.EnergyType.SOLAR,
+                            authdata={},
+                        )
+                        for loc in resp:
+                            if "gsp_id" in loc.metadata:
+                                gsp_id_map[int(loc.metadata["gsp_id"])] = loc
+                        log.info(f"Populated GSP ID map with {len(gsp_id_map)} entries")
+
+                    break
+                except Exception:
+                    log.warning(
+                        f"Failed to initialise dataplatform (attempt {attempt}/5)",
+                    )
+                    if attempt == 5:
+                        raise
+                    await asyncio.sleep(5)
 
         case _ as backend_type:
             raise ValueError(f"Unknown backend: {backend_type}")

